### PR TITLE
WIP: Port A* and floodFill from Neatness

### PR DESCRIPTION
**Pull request by jackunion**

### CODE REVIEW

**Comment by jackunion** [comment]
Do you want to keep `updateValue`, `updateFunction`, 'walkFunction' etc or do you want the "classic" versions?

There are some issues with types, will look into that

A WIP for now :)

**Comment by mvasilkov** [comment]
Hi, thanks for the PR! I wanted to take another look at those functions really, they can probably use some refactoring as stuff I write for js13k is usually quick and dirty and full of errors. :)

May I ask if you're using / planning to use natlib for something? And if yes, how can I assist with that.

**Comment by jackunion** [comment]
>  I wanted to take another look at those functions really, they can probably use some refactoring as stuff I write for js13k is usually quick and dirty and full of errors

Well, doing some refactoring was the idea, hence the question about extra arguments and your thoughts / plans for these functions as part of natlib. I pretty much copypasted them from neatness and slapped the good old WIP on this PR anyway.

> May I ask if you're using / planning to use natlib for something?
Not yet, but maybe in the future. Just got laid off and thought I'd do something meaningful with the free time I got and return some favours for old times sake :)

**Comment by mvasilkov** [comment]
> Just got laid off

So sorry to hear that!

**Comment by jackunion** [comment]
@mvasilkov A* should be good to go, I'll take a final look later to see if there's anything I don't like, but if you have time you can start reviewing the changes and tell me if you want anything changed / fixed / added / removed.

Also - do you prefer having a separate README in a folder or just a comment inside the file? I'll write some usage notes too.

**Comment by mvasilkov** [comment]
Hi! Sorry for slow replies.

> you can start reviewing the changes and tell me if you want anything changed

It's easier to show than tell :) But really I was thinking that

* Technically `A*` isn't even grid based, so a good implementation of the algo will be generic graph-based (and will support different costs)
* But then again I don't see myself using it on something other than a grid, so might as well stay grid-based
* However it absolutely shouldn't receive the grid itself! Because the algo still doesn't know nothing about how I want my levels to work (for example boundary check is was useful for The Neatness, but you can't make a circular level with it).

So yeah I'm a little conflicted myself with regards to how this should be. I'll work on it a bit and I'll post the results.

> wasn't called for the starting pixel

This is intentional — DDA also does that. The idea is, you're already in that location, so walking there is a no-op. (This also makes going from A to B to C to A consistent, we don't visit the same location twice.)

File: typescript/algorithms/astar/utils.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/** Manhattan distance on a square grid */
+export const manhattan = (
+  x0: number,
+  x1: number,
+  y0: number,
+  y1: number
+): number => {
+  return Math.abs(x0 - x1) + Math.abs(y0 - y1)
+}
```
Comment (mvasilkov) [inline typescript/algorithms/astar/utils.ts:52]: Regarding code review: e.g. this function being standalone and not a method of Vec2 is absolutely my mistake. I just write stuff to the file that happened to be open at the time. :)

Comment (jackunion) [reply typescript/algorithms/astar/utils.ts:52]: Maybe put it in some kind of `utils` folder along with euclidean distance and then extend `Vec2`?

**Comment by mvasilkov** [comment]
I'll push something today which is my vision for it. Just need to see if it works correctly first.

**Comment by mvasilkov** [comment]
Here, this is my take: https://github.com/mvasilkov/natlib/commit/2fa5d508b3ce582e01fb04de77578ae850071774#diff-b2c96a375d2e88f63e621031a6aeaaa0d3bfc6bb2b9fa662ec4d69fb8d1cdc62R1

The idea is, `GetAdjacent` is a generator function (`function*`) that yields tuples of `[x, y]`. This is not the cleanest way (that would be working with `Vec2` throughout) but I think it's good enough? Still way more memory allocations than I'm comfortable with.

**Comment by jackunion** [comment]
Closing this request since you've already ported both A* and floodFill :)

Any help you might need on other issues?

**Comment by mvasilkov** [comment]
Hey hey, yeah thanks to you I had just enough motivation to finally get this done :)

I think the most problematic thing right now is the lack of examples (and docs, but that's a whole another issue). The examples live here: https://github.com/mvasilkov/natlab

Usage on that repo, until I make a proper dev server:

1. `git clone`
2. (Optional) set up a Python virtual env
3. `yarn`
4. (Optional) link natlib from the master branch, e.g. `yarn link natlib`
5. `pip install -r requirements.txt`
6. `python build/build.py` this builds html pages from templates
7. `tsc -w`
8. In other tab: run any web server. E.g. `python3 -m http.server`

Reasonably complete demos are:
1. prng32
2. easing_functions
3. verlet_basic
4. dda

Here's some screens of what you should see:
Screen Shot 2023-02-02 at 16 38 12

Screen Shot 2023-02-02 at 16 38 50

Screen Shot 2023-02-02 at 16 39 03

**Comment by mvasilkov** [comment]
As for which examples are missing: like, all of them! The endgame is to make a number of very simple games (I've started one, in the folder flappy_birb, but got nowhere with is at this point...)

Stuff like
- Very basic platformer (no graphics, just some colored squares)
    - Collision detection for that is still missing from natlib, I have it on my endless TODO list...
- Simple flappy bird clone
- Simple physics-based game
- Literally anything else that can make use of natlib

**Pull request closed without merging**